### DISCUSSION TITLE
Add 2FA

### DISCUSSION THREAD

**Discussion by MagicLike** [umami-software/umami#1195] [Ideas]
For obvious security reasons (e.g. on a large scale), I would say that the option to use 2FA is a requirement.

**Comment by ThatOneCalculator** [comment]
Yes, please! I love using Umami but I feel this is something that's very necessary to add.

**Comment by boooong123** [comment]
Any updates on this feature request?

**Comment by samumatic** [comment]
2FA options like:
* WebAuthn
* OTP
* E-Mail

would be a secure basis.

**Comment by h0ek** [comment]
+1

**Comment by sudiptosarkar** [comment]
I just installed Umami, and I'm loving it already. But considering the fact this is a public facing service, the 2FA is a crucial missing ingredient.

Few ideas to begin with, on top of the options laid out by @samumatic:

* TOTP (Google Authenticator and stuff)
* Passkey (Fingerprints on phones and laptops, Face signin on Windows (works on Edge and Firefox for me, Chrome should work too))

**Comment by VMBindraban** [comment]
Any progress on this?

**Comment by mgntrn** [comment]
I'd like this too :)

**Comment by dewald-els** [comment]
We are looking at using Umami but unfortunately without 2FA we can not proceed. Is this something that is in the roadmap?

**Comment by rakurtz** [comment]
Very much appreciate this too. 2FA  is a must on a public facing site these days.

**Comment by 0xA1F1E** [comment]
I am baffled that this feature request for a basic security feature like 2FA is **3 YEARS** old! How has this not been implemented yet? Very much required for a service that is meant to be running publicly.

**Comment by h0ek** [comment]
I am happy to donate some 🪙  for this :) just let me how much you need 💸

**Comment by MitchellSimY** [comment]
Bump - I love to see this feature please. In the current self-hosted version of Umami, it appears that one could possibly brute force the UI. I would really love if there was email 2FA we can set up.

**Comment by sudiptosarkar** [comment]
@MitchellSimY, Email is more of a hassle than say, Google Authenticator.

We'd have to setup SMTP server connections.

**Reply by MitchellSimY** [reply]
I see - I wouldn't mind an Authenticator app at all either! That'd be cool

**Comment by guiyumin** [comment]
wow, still no 2fa

**Comment by drefrajo** [comment]
2fa would be very good to see indeed.

Especially TOTP and WebAuthn seem practical. (also, passkeys instead of username/password/2fa would be neat as well)

**Comment by boutterudy** [comment]
I like the idea of adding 2FA functionality to Umami, but in your opinion, **which approach would be best suited to Umami for implementing 2FA via an app** _(e.g. Google Authenticator, Proton Authenticator, etc.)_**?** 👀 

**A.** **Mandatory** for logging in, **enabled for everyone**
**B.** **Configurable** in settings _(toggle)_, **disabled by default**
**C.** **Configurable** in settings _(toggle)_, **enabled by default**
**D.** **Other approach**

Personally, given that this is an open-source application used by many people, I would tend to **initially** **allow the admin to configure it** _(toggle)_, and then, **at a later stage** _(once the system is fully in place, functional and tested)_, potentially **make it mandatory** to improve Umami’s security 💯 

Initially, I could see the 2FA system being **enabled by default for all new Umami installations**, and **disabled by default for existing installations** _(to avoid causing a mess for existing users)_ 🤔 

**What do you think? 👀**

**Reply by sudiptosarkar** [reply]
> I like the idea of adding 2FA functionality to Umami, but in your opinion, **which approach would be best suited to Umami for implementing 2FA via an app** _(e.g. Google Authenticator, Proton Authenticator, etc.)_**?** 👀
> 
> **A.** **Mandatory** for logging in, **enabled for everyone** **B.** **Configurable** in settings _(toggle)_, **disabled by default** **C.** **Configurable** in settings _(toggle)_, **enabled by default** **D.** **Other approach**
> 
> Personally, given that this is an open-source application used by many people, I would tend to **initially** **allow the admin to configure it** _(toggle)_, and then, **at a later stage** _(once the system is fully in place, functional and tested)_, potentially **make it mandatory** to improve Umami’s security 💯
> 
> Initially, I could see the 2FA system being **enabled by default for all new Umami installations**, and **disabled by default for existing installations** _(to avoid causing a mess for existing users)_ 🤔
> 
> **What do you think? 👀**

B., @boutterudy

**Reply by MitchellSimY** [reply]
Configurable for sure. 

I wouldn't mind built in support for an Authenticator app. I'd also be totally down for configurable SMTP values for email support too. Eg we provide our own SMTP values etc.

**Reply by boutterudy** [reply]
@sudiptosarkar @MitchellSimY Thank you very much for your replies! 🙏 

I’ve been working on a 2FA implementation with a more refined system than simply forcing it on everyone: administrators can make 2FA required for everyone; or just for certain teams; or just for certain users; it’s more flexible and seems well-suited Umami's features! 👏 

Please feel free to take a look at the pull request and give me some feedback on the code or security. I’ve done my best, but I’m 100% open to your feedback! 🫶 

Pull request: https://github.com/umami-software/umami/pull/4205

**Comment by kgrosvenor** [comment]
Pls test and merge the PR ❤️  it would bring such a huge security improvement

**Reply by neoighodaro** [reply]
+1, this is something i am considering for our product but without 2FA it will fail security audits 100%.
